README.md
Nimflux
=======

Nimflux is an InfluxDB API client library for Nim that can be used to `ping`,
`write`, `query`, or send custom `request`s. The `DataPoint` type is used to
easily create measurements for `write`, but it's also possible to send your own
Line Protocol string should the need arise.

```nim
import nimflux

var data = DataPoint(measurement: "temp")
data.addTag("loc", "home")
data.addField("ambient", 22.0)

var client = newInfluxClient("localhost", "nimtest")
var resp = client.write(data)
assert resp.code.toInfluxStatus == Ok
resp = client.query("select * from temp")
echo resp.body
```

Asynchronous actions are also supported through `AsyncInfluxClient`:

```nim
import asyncdispatch, nimflux

var data = DataPoint(measurement: "temp")
data.addTag("loc", "home")
data.addField("ambient", 22.0)

proc asyncProc(data: DataPoint): Future[AsyncResponse] {.async.} =
  var client = newAsyncInfluxClient("localhost", "nimtest")
  var resp = await client.write(data)
  assert resp.code.toInfluxStatus == Ok
  return await client.query("select * from temp")

echo asyncProc(data).waitFor().body.waitFor()
```


CHANGELOG.md
Changes
=======

[1.0.3] - 2021-11-23
--------------------

### Changed

* Fixed not being able to pass `SslContext` to `ssl=` in `newAsyncInfluxClient`.
* Fixed not being able to get a string representation of `AsyncInfluxClient`
  through `$`.
* Fixed `setBasicAuth` and `setTokenAuth` not working with `AsyncInfluxClient`.

[1.0.2] - 2021-07-01
--------------------

### Changed

* `close` was func instead of proc causing an issue with the `strictFunc` pragma
  has it may have possible side effects. `close` is now correctly proc.

[1.0.1] - 2021-06-29
--------------------

### Changed

* Fix value of tags being erroneously quoted.

[1.0.0] - 2021-06-28
--------------------

### Added

* Asynchronous functionality has been added through `AsyncInfluxClient`.
* Custom `SslContext` can now be used for HTTP actions.
* The module now exports some necessary parts from the httpclient module.

### Changed

* The Influx client now has a persistent HTTP client instead of creating a new
  instance for each request. The proc `close` has been added for closing the
  internal HTTP client.
* Requests no longer returns both the HTTP response and the "Influx status",
  just the HTTP response. The status can be created manually by using the
  exported `toInfluxStatus` proc.

[0.1.3] - 2021-04-08
--------------------

### Added

* Implemented `$` proc for InfluxClient.

[0.1.2] - 2021-03-11
--------------------

### Changed

* Fixed socket fd leaking due to not closing httpClient after requests. Leaking
  will still occurr on failed connections due to an upstream bug, see:
  https://github.com/nim-lang/Nim/issues/12381

[0.1.1] - 2021-03-01
--------------------

### Changed

* Fixed InfluxClient not being GC-safe.

[0.1.0] - 2021-02-27
--------------------

### Added

* Initial release


tests/t_client.nim
import asyncdispatch, httpclient, unittest

import nimflux

suite "InfluxClient":
  var client = newInfluxClient("localhost", "nimtest")

  test "ping":
    var resp = client.ping()
    check resp.code == Http204
    check resp.code.toInfluxStatus == InfluxStatus.Ok

  test "write incorrect":
    var resp = client.write("lorem ipsum")
    check resp.code == Http400
    check resp.code.toInfluxStatus == BadRequest

  test "write string":
    var resp = client.write("tm tf=1i")
    check resp.code == Http204
    check resp.code.toInfluxStatus == InfluxStatus.Ok

  test "write DataPoint":
    var d = DataPoint(measurement: "tm")
    d.addField("tf", 1)
    var resp = client.write(d)
    check resp.code == Http204
    check resp.code.toInfluxStatus == InfluxStatus.Ok

  test "write seq[DataPoint]":
    var d1 = DataPoint(measurement: "tm")
    var d2 = DataPoint(measurement: "tm")
    d1.addField("tf", 1)
    d2.addField("tf", 2)
    var resp = client.write(@[d1, d2])
    check resp.code == Http204
    check resp.code.toInfluxStatus == InfluxStatus.Ok

  test "query incorrect":
    var resp = client.query("")
    check resp.code == Http400
    check resp.code.toInfluxStatus == BadRequest

  test "query OK":
    var resp = client.query("select * from tm")
    check resp.code == Http200
    check resp.code.toInfluxStatus == InfluxStatus.Ok

suite "AsyncInfluxClient":
  var client = newAsyncInfluxClient("localhost", "nimtest")

  test "write DataPoint":
    var
      d1 = DataPoint(measurement: "tm")
    d1.addField("tf", 1)
    var future = client.write(d1)
    var resp = waitFor(future)
    check resp.code == Http204


tests/t_datapoint.nim
import unittest
import tables

import coverage

import nimflux

suite "DataPoint":

  test "addTag":
    let point = DataPoint()
    point.addTag("tag", "hello!")
    check point.tags.hasKey("tag") == true
    check point.tags["tag"] == "\"hello!\""

  test "addField string":
    let point = DataPoint()
    point.addField("fieldStr", "hello!")
    check point.fields.hasKey("fieldStr") == true
    check point.fields["fieldStr"] == "\"hello!\""

  test "addField int":
    let point = DataPoint()
    point.addField("fieldInt", 3)
    check point.fields.hasKey("fieldInt") == true
    check point.fields["fieldInt"] == "3i"

  test "addField float":
    let point = DataPoint()
    point.addField("fieldFloat", 1.0)
    check point.fields.hasKey("fieldFloat") == true
    check point.fields["fieldFloat"] == "1.0"

  test "addField bool":
    let point = DataPoint()
    point.addField("fieldTrue", true)
    point.addField("fieldFalse", false)
    check point.fields.hasKey("fieldTrue") == true
    check point.fields.hasKey("fieldFalse") == true
    check point.fields["fieldTrue"] == "t"
    check point.fields["fieldFalse"] == "f"

  test "`$` full":
    let point = DataPoint(measurement: "msr", tags: {"tag": "tval"}.toTable, timestamp: 1)
    point.addField("field", "fval")
    check $point == "msr,tag=tval field=\"fval\" 1"

  test "`$` sans tags":
    let point = DataPoint(measurement: "msr", timestamp: 1)
    point.addField("field", "fval")
    check $point == "msr field=\"fval\" 1"

  test "`$` sans timestamp":
    let point = DataPoint(measurement: "msr", tags: {"tag": "tval"}.toTable)
    point.addField("field", "fval")
    check $point == "msr,tag=tval field=\"fval\""

echo "Coverage by file: "
for fname, num in coveragePercentageByFile().pairs():
  echo fname, " ", num

echo "Total coverage: ", totalCoverage()


src/influx.nim
{.experimental: "strictFuncs".}
import base64, hashes, httpclient, httpcore, strutils, tables, uri

import coverage

type
  InfluxClient* = ref object
    host*: string
    port*: int
    ssl*: bool
    database*: string
    auth: string
    timeout*: int

  DataPoint* = ref object
    ## Representation of a single Influx data point.
    measurement*: string
    tags*: Table[string, string]
    fields*: Table[string, string]
    timestamp*: int64

  InfluxStatus* = enum
    Ok,
    BadRequest,
    Unauthorized,
    NotFound,
    RequestTooLarge,
    ServerError,
    UnknownError

func newInfluxClient*(host: string, database: string, port = 8086, ssl = false,
    timeout = -1): InfluxClient =
  ## Create a new InfluxClient instance for communicating with the InfluxDB API.
  ##
  ## ``host`` specifies the host to target.
  ##
  ## ``database`` specifies the default InfluxDB database to target.
  ##
  ## ``port`` specifies the port to target.
  ##
  ## ``ssl`` specifies the use of HTTPS communication.
  ##
  ## ``timeout`` specifies the number of milliseconds to allow before a
  ## ``TimeoutError`` is raised.
  InfluxClient(host: host, port: port, database: database, ssl: ssl,
      timeout: timeout)

func toInfluxStatus(code: HttpCode): InfluxStatus =
  if is2xx(code):
    OK
  elif is4xx(code):
    if code == Http400:
      BadRequest
    elif code == Http401:
      Unauthorized
    elif code == Http404:
      NotFound
    elif code == Http413:
      RequestTooLarge
    else:
      UnknownError
  elif is5xx(code):
    ServerError
  else:
    UnknownError

proc setBasicAuth*(i: InfluxClient, user: string, pwd: string) {.cov.} =
  ## Use Basic authentication.
  i.auth = "Basic " & encode(user & ":" & pwd)

proc setTokenAuth*(i: InfluxClient, token: string) {.cov.} =
  ## Use Token authentication.
  i.auth = "Token " & token

proc addField*(l: DataPoint, name: string, value: string) {.cov.} =
  ## Add a measurement field.
  l.fields[name] = "\"" & value & "\""

proc addField*(l: DataPoint, name: string, value: int) {.cov.} =
  l.fields[name] = $value & "i"

proc addField*(l: DataPoint, name: string, value: float) {.cov.} =
  l.fields[name] = $value

proc addField*(l: DataPoint, name: string, value: bool) {.cov.} =
  l.fields[name] = case value:
    of true:
      "t"
    of false:
      "f"

func `$`*(l: DataPoint): string =
  result = l.measurement
  for key, val in l.tags:
    result.add("," & key & "=" & val)
  result.add(" ")
  var fields = newSeq[string]()
  for key, val in l.fields:
    fields.add(key & "=" & val)
  result.add(fields.join(","))
  if l.timestamp != 0:
    result.add(" " & $l.timestamp)

proc request*(i: InfluxClient, endpoint: string, httpMethod = HttpGet,
    data = "", queryString: seq[(string, string)] = @[]): (Response,
        InfluxStatus) {.cov.} =
  ## Send request to Influx using connection values from InfluxClient directed
  ## at the specified InfluxDB API ``endpoint`` using the method specified by
  ## ``httpMethod``.
  var hostUri = initUri()
  if i.ssl:
    hostUri.scheme = "https"
  else:
    hostUri.scheme = "http"
  hostUri.hostname = i.host
  hostUri.port = $i.port
  hostUri.path = endpoint
  hostUri.query = encodeQuery(queryString)
  let client = newHttpClient(timeout = i.timeout)
  client.headers = newHttpHeaders()
  if i.auth != "":
    client.headers["Authorization"] = i.auth
  let r = client.post($hostUri, data)
  (r, r.code.toInfluxStatus())

proc ping*(i: InfluxClient): InfluxStatus =
  ## Ping InfluxDB to check instance status.
  i.request("/ping", HttpGet)[1]

proc query*(i: InfluxClient, q: string, database = "", chunked = false,
    chunkSize = 10000, epoch = "ns", pretty = false): (Response,
        InfluxStatus) {.cov.} =
  ## Query InfluxDB using InfluxQL. HTTP method is automatically determined by
  ## the query type in ``q``.
  ##
  ## ``q`` specifies the InfluxQL query to execute.
  ##
  ## ``database`` specifies the target database. This overrides the
  ## InfluxClient setting which would otherwise be used.
  ##
  ## ``chunked`` instructs the server to return points in streamed batches
  ## (chunks) instead of in a single response.
  ##
  ## ``chunkSize`` specifies the number of points constituting a chunk.
  ##
  ## ``epoch`` specifies the precision of the timestamps returned by the query.
  ##
  ## ``pretty`` instructs the server to pretty-print returned JSON.
  var querySeq = @[("q", q),
                   ("epoch", epoch),
                   ("pretty", $pretty)]
  if chunked:
    querySeq.add(("chunked", $chunked))
  if database != "":
    querySeq.add(("db", database))
  elif i.database != "":
    querySeq.add(("db", i.database))
  var meth: HttpMethod
  if q.toLowerAscii.startsWith("select") or q.toLowerAscii.startsWith("show"):
    meth = HttpGet
  else:
    meth = HttpPost
  i.request("/query", meth, queryString = querySeq)

proc write*(i: InfluxClient, data: string, database: string): InfluxStatus {.cov.} =
  ## Write data points to InfluxDB using Line Protocol.
  var db: string
  if database != "":
    db = database
  elif i.database != "":
    db = i.database
  else:
    return BadRequest
  i.request("/write", HttpPost, data, @[("db", db)])[1]

proc write*(i: InfluxClient, data: seq[DataPoint],
    database = ""): InfluxStatus =
  i.write(data.join("\n"), database)

proc write*(i: InfluxClient, data: DataPoint, database = ""): InfluxStatus =
  i.write($data, database)


src/nimflux.nim
# Copyright (c) 2021 Tobias Dély
#
# This software is licensed under the MIT license.
# See the file "LICENSE", included in this distribution, for details about the
# copyright.

## This module implements an InfluxDB API client.
##
## .. code-block:: Nim
##   import nimflux
##
##   var data = DataPoint(measurement: "temp")
##   data.addTag("loc", "home")
##   data.addField("ambient", 22.0)
##
##   var client = newInfluxClient("localhost", "nimtest")
##   var resp = client.write(data)
##   assert resp.code.toInfluxStatus == Ok
##   resp = client.query("select * from temp")
##   echo resp.body
##
## It also supports asynchronous actions through ``AsyncInfluxClient``:
##
## .. code-block:: Nim
##   import asyncdispatch, nimflux
##
##   var data = DataPoint(measurement: "temp")
##   data.addTag("loc", "home")
##   data.addField("ambient", 22.0)
##
##   proc asyncProc(data: DataPoint): Future[AsyncResponse] {.async.} =
##     var client = newAsyncInfluxClient("localhost", "nimtest")
##     var resp = await client.write(data)
##     assert resp.code.toInfluxStatus == Ok
##     return await client.query("select * from temp")
##
##   echo asyncProc(data).waitFor().body.waitFor()
##

{.experimental: "strictFuncs".}
import asyncdispatch, base64, hashes, httpclient, httpcore, net, strutils,
       tables, uri

export AsyncResponse, Response, body, code

type
  InfluxClientBase*[ClientType] = ref object
    httpClient: ClientType
    host*: string
    port*: int
    ssl*: bool
    database*: string
    auth: string

  InfluxClient* = InfluxClientBase[HttpClient]

  AsyncInfluxClient* = InfluxClientBase[AsyncHttpClient]

  DataPoint* = ref object
    ## Representation of a single Influx data point.
    measurement*: string
    tags*: Table[string, string]
    fields*: Table[string, string]
    timestamp*: int64

  InfluxStatus* {.pure.} = enum
    Ok,
    BadRequest,
    Unauthorized,
    NotFound,
    RequestTooLarge,
    ServerError,
    UnknownError

func `$`*(client: InfluxClient | AsyncInfluxClient): string =
  var hostUri = initUri()
  if client.ssl:
    hostUri.scheme = "https"
  else:
    hostUri.scheme = "http"
  hostUri.hostname = client.host
  hostUri.port = $client.port
  hostUri.query = encodeQuery(@[("db", client.database)])
  $hostUri

proc newInfluxClient*(host: string, database: string, port = 8086,
                      ssl: SslContext | bool = false, timeout = -1):
                      InfluxClient =
  ## Create a new InfluxClient instance for communicating with the InfluxDB API.
  ##
  ## ``host`` specifies the host to target.
  ##
  ## ``database`` specifies the default InfluxDB database to target.
  ##
  ## ``port`` specifies the port to target.
  ##
  ## ``ssl`` specifies the use of HTTPS communication. If value is a SslContext
  ## then that will be used in any HTTPS connection.
  ##
  ## ``timeout`` specifies the number of milliseconds to allow before a
  ## ``TimeoutError`` is raised.
  new result
  result.host = host
  result.port = port
  result.database = database
  when ssl is bool:
    result.ssl = ssl
    result.httpClient = newHttpClient(timeout = timeout)
  else:
    result.ssl = true
    result.httpClient = newHttpClient(timeout = timeout, sslContext = ssl)

proc newAsyncInfluxClient*(host: string, database: string, port = 8086,
                           ssl: SslContext | bool = false): AsyncInfluxClient =
  ## Create a new AsyncInfluxClient instance for communicating with the
  ## InfluxDB API.
  ##
  ## ``host`` specifies the host to target.
  ##
  ## ``database`` specifies the default InfluxDB database to target.
  ##
  ## ``port`` specifies the port to target.
  ##
  ## ``ssl`` specifies the use of HTTPS communication. If value is a SslContext
  ## then that will be used in any HTTPS connection.
  ##
  ## ``timeout`` specifies the number of milliseconds to allow before a
  ## ``TimeoutError`` is raised.
  new result
  result.host = host
  result.port = port
  result.database = database
  when ssl is bool:
    result.ssl = ssl
    result.httpClient = newAsyncHttpClient()
  else:
    result.ssl = true
    result.httpClient = newAsyncHttpClient(sslContext = ssl)

proc close*(client: InfluxClient | AsyncInfluxClient) =
  ## Close any HTTP connection used by the ``client``.
  client.httpClient.close()

func toInfluxStatus*(code: HttpCode): InfluxStatus =
  ## Get InfluxStatus from HTTP response code.
  if is2xx(code):
    OK
  elif is4xx(code):
    if code == Http400:
      BadRequest
    elif code == Http401:
      Unauthorized
    elif code == Http404:
      NotFound
    elif code == Http413:
      RequestTooLarge
    else:
      UnknownError
  elif is5xx(code):
    ServerError
  else:
    UnknownError

proc setBasicAuth*(client: InfluxClient | AsyncInfluxClient, user: string, pwd: string) =
  ## Use Basic authentication.
  client.auth = "Basic " & encode(user & ":" & pwd)

proc setTokenAuth*(client: InfluxClient | AsyncInfluxClient, token: string) =
  ## Use Token authentication.
  client.auth = "Token " & token

proc addTag*(dp: DataPoint, name: string, value: string) =
  ## Add a measurement tag.
  dp.tags[name] = value

proc addField*(dp: DataPoint, name: string, value: string) =
  ## Add a measurement field.
  dp.fields[name] = "\"" & value & "\""

proc addField*(dp: DataPoint, name: string, value: int) =
  dp.fields[name] = $value & "i"

proc addField*(dp: DataPoint, name: string, value: float) =
  dp.fields[name] = $value

proc addField*(dp: DataPoint, name: string, value: bool) =
  dp.fields[name] = case value:
    of true:
      "t"
    of false:
      "f"

func `$`*(dp: DataPoint): string =
  result = dp.measurement
  for key, val in dp.tags:
    result.add("," & key & "=" & val)
  result.add(" ")
  var fields = newSeq[string]()
  for key, val in dp.fields:
    fields.add(key & "=" & val)
  result.add(fields.join(","))
  if dp.timestamp != 0:
    result.add(" " & $dp.timestamp)

proc request*(client: InfluxClient | AsyncInfluxClient, endpoint: string,
              httpMethod = HttpGet, data = "",
              queryString: seq[(string, string)] = @[]):
              Future[Response | AsyncResponse] {.multisync.} =
  ## Send request to Influx using connection values from ``client`` directed
  ## at the specified InfluxDB API ``endpoint`` using the method specified by
  ## ``httpMethod``.
  var hostUri = initUri()
  if client.ssl:
    hostUri.scheme = "https"
  else:
    hostUri.scheme = "http"
  hostUri.hostname = client.host
  hostUri.port = $client.port
  hostUri.path = endpoint
  hostUri.query = encodeQuery(queryString)
  client.httpClient.headers = newHttpHeaders()
  if client.auth != "":
    client.httpClient.headers["Authorization"] = client.auth
  return await client.httpClient.request($hostUri, httpMethod, data)

proc ping*(client: InfluxClient | AsyncInfluxClient):
           Future[Response | AsyncResponse] {.multisync.} =
  ## Ping InfluxDB to check instance status.
  return await client.request("/ping", HttpGet)

proc query*(client: InfluxClient | AsyncInfluxClient, q: string, database = "",
            chunked = false, chunkSize = 10000, epoch = "ns", pretty = false):
            Future[Response | AsyncResponse] {.multisync.} =
  ## Query InfluxDB using InfluxQL. HTTP method is automatically determined by
  ## the query type in ``q``.
  ##
  ## ``q`` specifies the InfluxQL query to execute.
  ##
  ## ``database`` specifies the target database. This overrides the
  ## InfluxClient setting which would otherwise be used.
  ##
  ## ``chunked`` instructs the server to return points in streamed batches
  ## (chunks) instead of in a single response.
  ##
  ## ``chunkSize`` specifies the number of points constituting a chunk.
  ##
  ## ``epoch`` specifies the precision of the timestamps returned by the query.
  ##
  ## ``pretty`` instructs the server to pretty-print returned JSON.
  var querySeq = @[("q", q),
                   ("epoch", epoch),
                   ("pretty", $pretty)]
  if chunked:
    querySeq.add(("chunked", $chunked))
  if database != "":
    querySeq.add(("db", database))
  elif client.database != "":
    querySeq.add(("db", client.database))
  var mthd: HttpMethod
  if q.toLowerAscii.startsWith("select") or q.toLowerAscii.startsWith("show"):
    mthd = HttpGet
  else:
    mthd = HttpPost
  return await client.request("/query", mthd, queryString = querySeq)

proc write*(client: InfluxClient | AsyncInfluxClient, data: string,
            database: string = ""): Future[Response | AsyncResponse]
            {.multisync.} =
  ## Write data points to InfluxDB using Line Protocol.
  var db: string
  if database != "":
    db = database
  elif client.database != "":
    db = client.database
  return await client.request("/write", HttpPost, data, @[("db", db)])

proc write*(client: InfluxClient | AsyncInfluxClient, data: seq[DataPoint],
            database = ""): Future[Response | AsyncResponse] {.multisync.} =
  return await client.write(data.join("\n"), database)

proc write*(client: InfluxClient | AsyncInfluxClient, data: DataPoint,
            database = ""): Future[Response | AsyncResponse] {.multisync.} =
  return await client.write($data, database)
